## 1) Import rollingsales_manhattan.csv as a DataFrame

In [27]:
import numpy as np 
import pandas as pd

df = pd.read_csv("rollingsales_manhattan.csv", skiprows=4)
df.dropna(how="all", inplace=True)
print(df)

BOROUGH              NEIGHBORHOOD    BUILDING CLASS CATEGORY  \
0          1.0             ALPHABET CITY    01 ONE FAMILY DWELLINGS   
1          1.0             ALPHABET CITY    01 ONE FAMILY DWELLINGS   
2          1.0             ALPHABET CITY    02 TWO FAMILY DWELLINGS   
3          1.0             ALPHABET CITY    02 TWO FAMILY DWELLINGS   
4          1.0             ALPHABET CITY    02 TWO FAMILY DWELLINGS   
...        ...                       ...                        ...   
17154      1.0  WASHINGTON HEIGHTS UPPER  31 COMMERCIAL VACANT LAND   
17155      1.0  WASHINGTON HEIGHTS UPPER  31 COMMERCIAL VACANT LAND   
17156      1.0  WASHINGTON HEIGHTS UPPER  31 COMMERCIAL VACANT LAND   
17157      1.0  WASHINGTON HEIGHTS UPPER  31 COMMERCIAL VACANT LAND   
17158      1.0  WASHINGTON HEIGHTS UPPER           44 CONDO PARKING   

      TAX CLASS AT PRESENT   BLOCK     LOT  EASE-MENT  \
0                        1   376.0    43.0        NaN   
1                        1   400.0    19

The first 4 rows of the file only contained markup/description for the file, and the actual data started after the first 4 rows. In order to have read_csv parse the column and row labels correctly, the first 4 rows had to be omitted from the read_csv.

## 2) Display columns and row samples


In [28]:
print(df.columns)

print(df.head(5))

print(df.sample(n=5))

print(df.tail(5))

Index(['BOROUGH', 'NEIGHBORHOOD', 'BUILDING CLASS CATEGORY',
       'TAX CLASS AT PRESENT', 'BLOCK', 'LOT', 'EASE-MENT',
       'BUILDING CLASS AT PRESENT', 'ADDRESS', 'APARTMENT NUMBER', 'ZIP CODE',
       'RESIDENTIAL UNITS', 'COMMERCIAL UNITS', 'TOTAL UNITS',
       'LAND SQUARE FEET', 'GROSS SQUARE FEET', 'YEAR BUILT',
       'TAX CLASS AT TIME OF SALE', 'BUILDING CLASS AT TIME OF SALE',
       ' SALE PRICE ', 'SALE DATE'],
      dtype='object')
   BOROUGH   NEIGHBORHOOD  BUILDING CLASS CATEGORY TAX CLASS AT PRESENT  \
0      1.0  ALPHABET CITY  01 ONE FAMILY DWELLINGS                    1   
1      1.0  ALPHABET CITY  01 ONE FAMILY DWELLINGS                    1   
2      1.0  ALPHABET CITY  02 TWO FAMILY DWELLINGS                    1   
3      1.0  ALPHABET CITY  02 TWO FAMILY DWELLINGS                    1   
4      1.0  ALPHABET CITY  02 TWO FAMILY DWELLINGS                    1   

   BLOCK   LOT  EASE-MENT BUILDING CLASS AT PRESENT              ADDRESS  \
0  376.0  43.0     

## 3) Describe the rows and data types


In [29]:
print(df.columns)

print(df.dtypes)

print(df.count())

Index(['BOROUGH', 'NEIGHBORHOOD', 'BUILDING CLASS CATEGORY',
       'TAX CLASS AT PRESENT', 'BLOCK', 'LOT', 'EASE-MENT',
       'BUILDING CLASS AT PRESENT', 'ADDRESS', 'APARTMENT NUMBER', 'ZIP CODE',
       'RESIDENTIAL UNITS', 'COMMERCIAL UNITS', 'TOTAL UNITS',
       'LAND SQUARE FEET', 'GROSS SQUARE FEET', 'YEAR BUILT',
       'TAX CLASS AT TIME OF SALE', 'BUILDING CLASS AT TIME OF SALE',
       ' SALE PRICE ', 'SALE DATE'],
      dtype='object')
BOROUGH                           float64
NEIGHBORHOOD                       object
BUILDING CLASS CATEGORY            object
TAX CLASS AT PRESENT               object
BLOCK                             float64
LOT                               float64
EASE-MENT                         float64
BUILDING CLASS AT PRESENT          object
ADDRESS                            object
APARTMENT NUMBER                   object
ZIP CODE                          float64
RESIDENTIAL UNITS                 float64
COMMERCIAL UNITS                  float64


Columns with "wrong" type:

    - NEIGHBORHOOD: str
    - BUILDING CLASS CATEGORY: str
    - TAX CLASS AT PRESENT: str
    - BUILDING CLASS AT PRESENT: str
    - ADDRESS: str
    - APARTMENT NUMBER: str
    - LAND SQUARE FEET: float64
    - GROSS SQUARE FEET: float64
    - BUILDING CLASS AT TIME OF SALE: str
    - SALE PRICE: float64

In [30]:
# pd.options.display.float_format = '{:,.0f}'.format

# df[['NEIGHBORHOOD', 'BUILDING CLASS CATEGORY', 'TAX CLASS AT PRESENT', 'BUILDING CLASS AT PRESENT', 'ADDRESS', 'BUILDING CLASS AT TIME OF SALE']] = df[['NEIGHBORHOOD', 'BUILDING CLASS CATEGORY', 'TAX CLASS AT PRESENT', 'BUILDING CLASS AT PRESENT', 'ADDRESS', 'BUILDING CLASS AT TIME OF SALE']].applymap(str)

df['LAND SQUARE FEET'] = df['LAND SQUARE FEET'].str.replace(',', '').apply(float)
df['GROSS SQUARE FEET'] = df['GROSS SQUARE FEET'].str.replace(',', '').apply(float)
df[' SALE PRICE '] = df[' SALE PRICE '].str.replace(',', '').apply(float)
df['YEAR BUILT'] = df['YEAR BUILT'].astype(str).str.replace(',', '').apply(float)



print(df)
print(df.dtypes)

BOROUGH              NEIGHBORHOOD    BUILDING CLASS CATEGORY  \
0          1.0             ALPHABET CITY    01 ONE FAMILY DWELLINGS   
1          1.0             ALPHABET CITY    01 ONE FAMILY DWELLINGS   
2          1.0             ALPHABET CITY    02 TWO FAMILY DWELLINGS   
3          1.0             ALPHABET CITY    02 TWO FAMILY DWELLINGS   
4          1.0             ALPHABET CITY    02 TWO FAMILY DWELLINGS   
...        ...                       ...                        ...   
17154      1.0  WASHINGTON HEIGHTS UPPER  31 COMMERCIAL VACANT LAND   
17155      1.0  WASHINGTON HEIGHTS UPPER  31 COMMERCIAL VACANT LAND   
17156      1.0  WASHINGTON HEIGHTS UPPER  31 COMMERCIAL VACANT LAND   
17157      1.0  WASHINGTON HEIGHTS UPPER  31 COMMERCIAL VACANT LAND   
17158      1.0  WASHINGTON HEIGHTS UPPER           44 CONDO PARKING   

      TAX CLASS AT PRESENT   BLOCK     LOT  EASE-MENT  \
0                        1   376.0    43.0        NaN   
1                        1   400.0    19

## 4) Initial column (or row) clean-up


In [31]:
print(df.iloc[[9, 127], [8, 9]])

df.drop(columns=['BOROUGH', 'EASE-MENT', 'APARTMENT NUMBER'], inplace=True)

df.rename(columns={' SALE PRICE ': 'SALE PRICE'}, inplace=True)

print(df.columns)


ADDRESS APARTMENT NUMBER
9    275 EAST 7TH STREET, GD              NaN
127    259 EAST 7 STREET, 1B               1B
Index(['NEIGHBORHOOD', 'BUILDING CLASS CATEGORY', 'TAX CLASS AT PRESENT',
       'BLOCK', 'LOT', 'BUILDING CLASS AT PRESENT', 'ADDRESS', 'ZIP CODE',
       'RESIDENTIAL UNITS', 'COMMERCIAL UNITS', 'TOTAL UNITS',
       'LAND SQUARE FEET', 'GROSS SQUARE FEET', 'YEAR BUILT',
       'TAX CLASS AT TIME OF SALE', 'BUILDING CLASS AT TIME OF SALE',
       'SALE PRICE', 'SALE DATE'],
      dtype='object')


These columns should be removed because they are redundant: EASE-MENT only contained NA values as shown through the count() method, BOROUGH is 1 for all entries as the dataset is only for Manhattan, and APARTMENT NUMBER is already included in all of the ADDRESS fields if present, and some APARTMENT NUMBER fields are empty even though their respective ADDRESS field contained an apartment number. For example, row 9 lists the apartment number in the ADDRESS field, but the APARTMENT NUMBER field is empty, and row 127 contains the apartment number in both fields.

This column should be renamed because the extra spaces in the name make it inconsistent with the rest of the column names.

## 5) Determine the top three neighborhoods that had the most properties sold


In [32]:
df['NEIGHBORHOOD'].value_counts()

UPPER EAST SIDE (59-79)      1696
UPPER EAST SIDE (79-96)      1629
UPPER WEST SIDE (59-79)      1499
MIDTOWN EAST                 1006
UPPER WEST SIDE (79-96)       882
MIDTOWN WEST                  800
HARLEM-CENTRAL                760
GRAMERCY                      700
CHELSEA                       637
GREENWICH VILLAGE-CENTRAL     609
LOWER EAST SIDE               598
GREENWICH VILLAGE-WEST        557
MURRAY HILL                   551
TRIBECA                       512
SOHO                          407
UPPER WEST SIDE (96-116)      376
FLATIRON                      319
FINANCIAL                     314
KIPS BAY                      293
MIDTOWN CBD                   242
CLINTON                       242
MANHATTAN VALLEY              230
WASHINGTON HEIGHTS UPPER      228
HARLEM-EAST                   225
FASHION                       198
CIVIC CENTER                  179
WASHINGTON HEIGHTS LOWER      178
JAVITS CENTER                 166
CHINATOWN                     164
EAST VILLAGE  

Top 3 neighborhoods with number of sales: 

1) UPPER EAST SIDE (59-79)      1696
2) UPPER EAST SIDE (79-96)      1629
3) UPPER WEST SIDE (59-79)      1499

As each row documents a single property sale, the frequency of each neighborhood listed in the NEIGHBORHOOD column is the number of sales in that neighborhood, so a simple value_counts() call on the NEIGHBORHOOD column displays the number of sales in each neighborhood.

## 6) Describe the kind of buildings that were sold


In [33]:
# df['BUILDING CLASS CATEGORY'].value_counts()
# df['BUILDING CLASS AT TIME OF SALE'].value_counts()
# df['BUILDING CLASS AT TIME OF SALE'].unique()

building_types = pd.Series([0, 0, 0, 0], index=['One family homes', 'Office buildings', 'Condominiums', "Other"])
# print(building_types)

def parse_building_type(code):
    if code[0] == 'A':
        building_types['One family homes'] += 1
    elif code[0] == 'O':
        building_types['Office buildings'] += 1
    elif code[0] == 'R':
        building_types['Condominiums'] += 1
    else:
        building_types['Other'] += 1

df['BUILDING CLASS AT TIME OF SALE'].apply(parse_building_type)
print(building_types)

One family homes     131
Office buildings     146
Condominiums        9120
Other               7762
dtype: int64



Steps:
- Read the documentation, which states that building class codes that start with 'A' are one family homes, 'O' are office buildings, and 'R' are condominiums.

- Created a Series called building_types to store the count of each building type.

- Created a function parse_building_type to be called on each element in BUILDING CLASS AT TIME OF SALE, which reads the first letter of the building class code and adds to the count of the respective building type in building_types.

- Displayed the count by printing building_types.

## 7) Calculate summary statistics for the prices of properties sold for all of Manhattan and for a couple of select neighborhoods


In [42]:
df['SALE PRICE'].describe()
print('Chinatown:\n')
print(df.loc[df['NEIGHBORHOOD'] == 'CHINATOWN']['SALE PRICE'].describe())
print('\nEast Village:\n')
print(df.loc[df['NEIGHBORHOOD'] == 'EAST VILLAGE']['SALE PRICE'].describe())

Chinatown:

count    1.640000e+02
mean     2.467500e+06
std      4.756622e+06
min      0.000000e+00
25%      0.000000e+00
50%      8.000000e+05
75%      2.218750e+06
max      3.000000e+07
Name: SALE PRICE, dtype: float64

East Village:

count    1.630000e+02
mean     2.919427e+06
std      7.319655e+06
min      0.000000e+00
25%      1.180250e+05
50%      8.700000e+05
75%      1.790000e+06
max      7.315000e+07
Name: SALE PRICE, dtype: float64


Summary analysis:
- Mean: East Village has a higher average sale price, with a mean of $2,919,427
- Median: East Village has a higher median sale price, with a median of $870,000
- 25%: East Village has a higher 25 percentile, $118,250
- 75%: Chinatown has a higher 75 percentile, $2,218,750
- Max: East Village has the higher max sale price, $73,150,000
- Min: Both Chinatown and East Village have a minimum sale price of $0

Steps:

- Called describe on SALE PRICE to get the summary statistics on sale price for the whole dataset.
- Called loc to find all entries for Chinatown and East Village, indexed for the sale price for those entries, and called describe on the result to retrieve summary statistics for the sale prices of the individual neighborhoods.

